### ISSUE TITLE
Incorrect implementation of consistent-with for unions

### ISSUE THREAD

**Issue by JelleZijlstra** [astral-sh/ty#424]
### Summary

```python
from typing import Any

def f(x: list[int]):
    g(x)
    
def g(x: list[bool | Any]):
    f(x)
```

ty reports errors on both of these (playground as of today):

```
    Argument to function `g` is incorrect: Expected `list[bool | Any]`, found `list[int]` (invalid-argument-type) [Ln 4, Col 7]
    Argument to function `f` is incorrect: Expected `list[int]`, found `list[bool | Any]` (invalid-argument-type) [Ln 7, Col 7]
```

I think this is wrong and both of these calls should be accepted:

- Invariant types are assignable if the type parameters are consistent with each other. (We don't say this explicitly in the spec, but I think that's how invariance should work.)
- Two gradual types are consistent with each other if there is some materialization that makes them equivalent.
- In `Any | bool`, we could materialize `Any` to `int`, which reduces to `int | bool` -> `int` (as `bool` is a subtype of `int`).
- Therefore, `Any | bool` and `int` are consistent.

Pyright also gets this wrong, though only in one direction (it accepts the `g(x)` call but not the `f(x)` call). Mypy gets it right.

**Comment by JelleZijlstra** [comment]
Another example where ty incorrectly reports an error in both directions:

```python
from typing import Any

def f(x: list[list[Any]]):
    g(x)
    
def g(x: list[list[int | str]]):
    f(x)
```

**Comment by JelleZijlstra** [comment]
And another one where ty incorrectly reports two errors:

```python
from typing import Any

def f(x: list[tuple[Any] | tuple[bool]]):
    g(x)

def g(x: list[tuple[int]]):
    f(x)
```

(This is a counterexample to an implementation technique I was considering, where I'd special-case unions containing Any.)
